File: scripts/fix-article-navigation.py
#!/usr/bin/env python3
"""
Fix Article Navigation: Language Switcher + Back-to-News Top Nav

This script ensures ALL news articles have:
1. A language switcher nav (14 languages) after <body>
2. An article-top-nav div with a localized back-to-news link before the article

It auto-discovers all articles in the news/ directory and processes them
idempotently — safe to run multiple times.

Usage:
    python3 scripts/fix-article-navigation.py
    python3 scripts/fix-article-navigation.py --dry-run
"""

import os
import re
import sys
from pathlib import Path
from collections import defaultdict

# ── Language configuration ───────────────────────────────────────────────
LANGUAGES = ["en", "sv", "da", "no", "fi", "de", "fr", "es", "nl", "ar", "he", "ja", "ko", "zh"]

LANG_DISPLAY = {
    "en": ("🇬🇧", "English"),
    "sv": ("🇸🇪", "Svenska"),
    "da": ("🇩🇰", "Dansk"),
    "no": ("🇳🇴", "Norsk"),
    "fi": ("🇫🇮", "Suomi"),
    "de": ("🇩🇪", "Deutsch"),
    "fr": ("🇫🇷", "Français"),
    "es": ("🇪🇸", "Español"),
    "nl": ("🇳🇱", "Nederlands"),
    "ar": ("🇸🇦", "العربية"),
    "he": ("🇮🇱", "עברית"),
    "ja": ("🇯🇵", "日本語"),
    "ko": ("🇰🇷", "한국어"),
    "zh": ("🇨🇳", "中文"),
}

LANG_SWITCHER_ARIA = {
    "en": "Language", "sv": "Språk", "da": "Sprog", "no": "Språk",
    "fi": "Kieli", "de": "Sprache", "fr": "Langue", "es": "Idioma",
    "nl": "Taal", "ar": "اللغة", "he": "שפה", "ja": "言語",
    "ko": "언어", "zh": "语言",
}

BACK_TO_NEWS = {
    "en": "Back to News", "sv": "Tillbaka till nyheter",
    "da": "Tilbage til nyheder", "no": "Tilbake til nyheter",
    "fi": "Takaisin uutisiin", "de": "Zurück zu Nachrichten",
    "fr": "Retour aux actualités", "es": "Volver a noticias",
    "nl": "Terug naar nieuws", "ar": "العودة إلى الأخبار",
    "he": "חזרה לחדשות", "ja": "ニュースに戻る",
    "ko": "뉴스로 돌아가기", "zh": "返回新闻",
}

# ── Helpers ───────────────────────────────────────────────────────────────

def extract_lang(filename):
    """Extract language code from article filename."""
    name = filename.replace('.html', '')
    for lang in LANGUAGES:
        if name.endswith(f'-{lang}'):
            return lang
    return None


def extract_base(filename):
    """Extract base slug (without lang suffix) from article filename."""
    name = filename.replace('.html', '')
    for lang in LANGUAGES:
        if name.endswith(f'-{lang}'):
            return name[:-len(f'-{lang}')]
    return None


def news_index_for(lang):
    """Return the news index filename for a language."""
    if lang == 'en':
        return 'index.html'
    return f'index_{lang}.html'


def generate_language_switcher(base_slug, current_lang):
    """Generate language switcher HTML matching template.ts output."""
    aria = LANG_SWITCHER_ARIA.get(current_lang, "Language")
    lines = [f'  <nav class="language-switcher" role="navigation" aria-label="{aria}">']
    for lang in LANGUAGES:
        flag, name = LANG_DISPLAY[lang]
        active = ' active' if lang == current_lang else ''
        aria_current = ' aria-current="page"' if lang == current_lang else ''
        lines.append(
            f'    <a href="{base_slug}-{lang}.html" class="lang-link{active}" '
            f'hreflang="{lang}"{aria_current}>{flag} {name}</a>'
        )
    lines.append('  </nav>')
    return '\n'.join(lines)


def generate_top_nav(lang):
    """Generate article-top-nav div matching template.ts output."""
    label = BACK_TO_NEWS.get(lang, BACK_TO_NEWS["en"])
    index = news_index_for(lang)
    return (
        f'\n<div class="article-top-nav">\n'
        f'  <a href="{index}" class="back-to-news">\n'
        f'    \u2190 {label}\n'
        f'  </a>\n'
        f'</div>\n'
    )


# ── Processing ────────────────────────────────────────────────────────────

def process_article(filepath, base_slug, lang, dry_run=False):
    """
    Ensure an article has both language-switcher and article-top-nav.
    Returns a tuple (added_switcher, added_topnav).
    """
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    original = content
    added_switcher = False
    added_topnav = False

    # ── 1. Language switcher ──────────────────────────────────────────
    has_switcher = 'language-switcher' in content
    if not has_switcher:
        switcher_html = generate_language_switcher(base_slug, lang)
        # Insert after <body> tag
        content = re.sub(
            r'(<body>)',
            r'\1\n' + switcher_html,
            content,
            count=1,
        )
        added_switcher = True
    else:
        # Update existing switcher to have all 14 languages
        new_switcher = generate_language_switcher(base_slug, lang)
        content = re.sub(
            r'<nav class="language-switcher"[^>]*>.*?</nav>',
            new_switcher,
            content,
            flags=re.DOTALL,
            count=1,
        )

    # ── 2. article-top-nav ────────────────────────────────────────────
    has_topnav = 'article-top-nav' in content
    if not has_topnav:
        top_nav_html = generate_top_nav(lang)
        # Insert before the article/div.news-article element
        # Try: before <article class="news-article"> or <div class="news-article">
        inserted = False

        # Pattern A: insert after closing </nav> of language-switcher
        if '</nav>' in content:
            # Find the LAST </nav> that belongs to the language-switcher
            # Insert top-nav right after it
            nav_pattern = r'(</nav>)(\s*)(<(?:article|div)\s+class="(?:news-article|container)")'
            match = re.search(nav_pattern, content, re.DOTALL)
            if match:
                content = content[:match.end(1)] + top_nav_html + match.group(2) + match.group(3) + content[match.end():]
                inserted = True

        # Pattern B: insert directly before <article class="news-article"> or <article class="container">
        if not inserted:
            article_pattern = r'(<(?:article|div)\s+class="(?:news-article|container)")'
            match = re.search(article_pattern, content)
            if match:
                content = content[:match.start()] + top_nav_html + '\n' + content[match.start():]
                inserted = True

        if inserted:
            added_topnav = True

    # ── Write if changed ──────────────────────────────────────────────
    if content != original and not dry_run:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)

    return added_switcher, added_topnav


def discover_articles(news_dir):
    """Discover all article files grouped by base slug."""
    articles = defaultdict(dict)  # base_slug -> {lang: filepath}
    for html_file in sorted(news_dir.glob('*.html')):
        if html_file.name.startswith('index'):
            continue
        lang = extract_lang(html_file.name)
        base = extract_base(html_file.name)
        if lang and base:
            articles[base][lang] = html_file
    return articles


def main():
    dry_run = '--dry-run' in sys.argv
    if dry_run:
        print("=== DRY RUN — no files will be modified ===\n")

    # Resolve news directory relative to this script
    script_dir = Path(__file__).resolve().parent
    news_dir = script_dir.parent / 'news'

    if not news_dir.exists():
        print(f"ERROR: news directory not found at {news_dir}")
        sys.exit(1)

    print("=== Fix Article Navigation ===")
    print(f"News directory: {news_dir}\n")

    articles = discover_articles(news_dir)
    print(f"Discovered {len(articles)} unique article slugs\n")

    total = 0
    switchers_added = 0
    topnavs_added = 0

    for base_slug in sorted(articles.keys()):
        lang_files = articles[base_slug]
        for lang in LANGUAGES:
            if lang not in lang_files:
                continue
            filepath = lang_files[lang]
            total += 1
            s, t = process_article(filepath, base_slug, lang, dry_run=dry_run)
            if s:
                switchers_added += 1
            if t:
                topnavs_added += 1

    print("=== Summary ===")
    print(f"Total files processed: {total}")
    print(f"Language switchers added/updated: {switchers_added}")
    print(f"Top nav (article-top-nav) added: {topnavs_added}")
    if dry_run:
        print("\n(Dry run — no files were modified)")
    print("\n✓ Done!")


if __name__ == '__main__':
    main()

File: scripts/fix-article-navigation.ts
#!/usr/bin/env -S npx tsx
/**
 * Fix Article Navigation: Language Switcher + Back-to-News Top Nav
 *
 * This script ensures ALL news articles have:
 * 1. A language switcher nav (14 languages) after <body>
 * 2. An article-top-nav div with a localized back-to-news link before the article
 *
 * It auto-discovers all articles in the news/ directory and processes them
 * idempotently — safe to run multiple times.
 *
 * Usage:
 *   npx tsx scripts/fix-article-navigation.ts
 *   npx tsx scripts/fix-article-navigation.ts --dry-run
 *
 * @author Hack23 AB
 * @license Apache-2.0
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import type { Language } from './types/language.js';
import { ALL_LANG_CODES, FOOTER_LABELS } from './article-template/constants.js';
import { getNewsIndexFilename, generateArticleLanguageSwitcher } from './article-template/helpers.js';

// ── Helpers ───────────────────────────────────────────────────────────────

function extractLang(filename: string): Language | null {
  const name = filename.replace(/\.html$/, '');
  for (const lang of ALL_LANG_CODES) {
    if (name.endsWith(`-${lang}`)) return lang;
  }
  return null;
}

function extractBase(filename: string): string | null {
  const name = filename.replace(/\.html$/, '');
  for (const lang of ALL_LANG_CODES) {
    if (name.endsWith(`-${lang}`)) return name.slice(0, -(lang.length + 1));
  }
  return null;
}

function generateTopNav(lang: Language): string {
  const label = FOOTER_LABELS[lang].backToNews;
  const index = getNewsIndexFilename(lang);
  return `\n<div class="article-top-nav">\n  <a href="${index}" class="back-to-news">\n    ← ${label}\n  </a>\n</div>\n`;
}

// ── Processing ────────────────────────────────────────────────────────────

export interface ProcessResult {
  addedSwitcher: boolean;
  addedTopnav: boolean;
  fixedTopnav: boolean;
}

/**
 * Transform article HTML string: ensures language-switcher and article-top-nav
 * with a back-to-news link are present. Idempotent — safe to call multiple times.
 *
 * @param content   Original HTML string
 * @param baseSlug  Article base slug without directory prefix (e.g. "2026-01-01-article")
 * @param lang      Language code for this variant
 * @returns Updated HTML string and flags indicating what changed
 */
export function transformContent(
  content: string,
  baseSlug: string,
  lang: Language,
): { content: string } & ProcessResult {
  let result = content;
  let addedSwitcher = false;
  let addedTopnav = false;
  let fixedTopnav = false;

  // ── 1. Language switcher ──────────────────────────────────────────
  const hasSwitcher = result.includes('language-switcher');
  if (!hasSwitcher) {
    const switcherHtml = generateArticleLanguageSwitcher(baseSlug, lang);
    // Prefer inserting AFTER the skip-link so it remains the first focusable element.
    const skipLinkPattern = /(<a[^>]*class="skip-link"[^>]*>[\s\S]*?<\/a>)/;
    if (skipLinkPattern.test(result)) {
      result = result.replace(skipLinkPattern, `$1\n${switcherHtml}`);
    } else {
      result = result.replace(/(<body[^>]*>)/, `$1\n${switcherHtml}`);
    }
    addedSwitcher = true;
  } else {
    // Update existing switcher to have all 14 languages.
    // Use [^\S\n]* (spaces/tabs, not newlines) to consume any indentation before <nav>,
    // so the replacement string's own leading spaces don't accumulate on repeated runs.
    const newSwitcher = generateArticleLanguageSwitcher(baseSlug, lang);
    result = result.replace(/[^\S\n]*<nav class="language-switcher"[^>]*>[\s\S]*?<\/nav>/, newSwitcher);
  }

  // ── 2. article-top-nav ────────────────────────────────────────────
  const hasTopnav = result.includes('article-top-nav');

  // If top-nav exists but is missing back-to-news link, replace it
  if (hasTopnav) {
    const topNavHasBackLink =
      /<div class="article-top-nav">[\s\S]*?class="back-to-news"[\s\S]*?<\/div>/.test(result);
    if (!topNavHasBackLink) {
      result = result.replace(
        /<div class="article-top-nav">[\s\S]*?<\/div>/,
        generateTopNav(lang).trim(),
      );
      fixedTopnav = true;
    }
  }

  if (!hasTopnav) {
    const topNavHtml = generateTopNav(lang);
    let inserted = false;

    // Pattern A: insert after closing </nav> of language-switcher, before article/div.news-article
    if (result.includes('</nav>')) {
      const navPattern = /((<\/nav>)([\s]*)(<(?:article|div)\s+class="(?:news-article|container)"))/s;
      const match = navPattern.exec(result);
      if (match) {
        const endOfNav = match.index + match[2].length;
        const whitespace = match[3];
        const articleTag = match[4];
        const afterFull = result.slice(match.index + match[0].length);
        result = result.slice(0, endOfNav) + topNavHtml + whitespace + articleTag + afterFull;
        inserted = true;
      }
    }

    // Pattern B: insert directly before <article class="news-article"> or <div class="container">
    if (!inserted) {
      const articlePattern = /(<(?:article|div)\s+class="(?:news-article|container)")/;
      const match = articlePattern.exec(result);
      if (match) {
        result = result.slice(0, match.index) + topNavHtml + '\n' + result.slice(match.index);
        inserted = true;
      }
    }

    if (inserted) addedTopnav = true;
  }

  return { content: result, addedSwitcher, addedTopnav, fixedTopnav };
}

function processArticle(filepath: string, baseSlug: string, lang: Language, dryRun: boolean): ProcessResult {
  const original = fs.readFileSync(filepath, 'utf-8');
  const { content, addedSwitcher, addedTopnav, fixedTopnav } = transformContent(original, baseSlug, lang);

  // ── Write if changed ──────────────────────────────────────────────
  if (content !== original && !dryRun) {
    fs.writeFileSync(filepath, content, 'utf-8');
  }

  return { addedSwitcher, addedTopnav, fixedTopnav };
}

interface ArticleMap {
  [baseSlug: string]: Partial<Record<Language, string>>;
}

function discoverArticles(newsDir: string): ArticleMap {
  const articles: ArticleMap = {};
  const files = fs.readdirSync(newsDir).sort();
  for (const name of files) {
    if (!name.endsWith('.html') || name.startsWith('index')) continue;
    const lang = extractLang(name);
    const base = extractBase(name);
    if (lang && base) {
      if (!articles[base]) articles[base] = {};
      articles[base][lang] = path.join(newsDir, name);
    }
  }
  return articles;
}

function main(): void {
  const dryRun = process.argv.includes('--dry-run');
  if (dryRun) {
    console.log('=== DRY RUN — no files will be modified ===\n');
  }

  const scriptDir = path.dirname(fileURLToPath(import.meta.url));
  const newsDir = path.resolve(scriptDir, '..', 'news');

  if (!fs.existsSync(newsDir)) {
    console.error(`ERROR: news directory not found at ${newsDir}`);
    process.exit(1);
  }

  console.log('=== Fix Article Navigation ===');
  console.log(`News directory: ${newsDir}\n`);

  const articles = discoverArticles(newsDir);
  const slugs = Object.keys(articles).sort();
  console.log(`Discovered ${slugs.length} unique article slugs\n`);

  let total = 0;
  let switchersAdded = 0;
  let topnavsAdded = 0;
  let topnavsFixed = 0;

  for (const baseSlug of slugs) {
    const langFiles = articles[baseSlug];
    for (const lang of ALL_LANG_CODES) {
      const filepath = langFiles[lang];
      if (!filepath) continue;
      total++;
      const { addedSwitcher, addedTopnav, fixedTopnav } = processArticle(filepath, baseSlug, lang, dryRun);
      if (addedSwitcher) switchersAdded++;
      if (addedTopnav) topnavsAdded++;
      if (fixedTopnav) topnavsFixed++;
    }
  }

  console.log('=== Summary ===');
  console.log(`Total files processed: ${total}`);
  console.log(`Language switchers added: ${switchersAdded}`);
  console.log(`Top nav (article-top-nav) added: ${topnavsAdded}`);
  console.log(`Top nav fixed (missing back-to-news link): ${topnavsFixed}`);
  if (dryRun) {
    console.log('\n(Dry run — no files were modified)');
  }
  console.log('\n✓ Done!');
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}
